### DISCUSSION TITLE
Inversion of Control for controllers

### DISCUSSION THREAD

**Discussion by koubas** [oas-tools/oas-tools.github.io#10] [Q&A] [answered]
First, I'm glad that this project starts to show signs of life again, thanks! I'm pretty new to oas-tools and modern js/ts development. I don't miss a full blown DI containers and so on but I'm wondering, how do I achieve basic Inversion of Control for OAS Tools controllers? From all the examples I see it wants me to import the dependencies (services, configs, ...) in the controller module, but I want them to be injectable during creation so I can mock them in tests. Can this be achieved?

**Comment by alesancor1** [comment, marked as answer]
Unfortunately, DI is not possible in OAS Tools currently. However, since the framework is flexible enough, you could implement a basic DI that can serve your purpose. To do that, you need to create a new external module that replaces the router middleware and add it to the OAS Tools chain through `oasTools.use( )`. You could then inject your dependencies through this new module's configuration:

1. Create a new file for the new module and copy the [router middleware](https://github.com/oas-tools/oas-tools/blob/main/src/middleware/native/oas-router.js) code inside:
```js
// newRouter.js
import { OASBase, errors, logger } from "@oas-tools/commons";
import { commons } from "../../utils/index.js";
import { pathToFileURL } from "url";

export class OASRouter extends OASBase {

  constructor(oasFile, middleware) {
    super(oasFile, middleware);
  }

  static async initialize(oasFile, config) {
    const controllers = await OASRouter.#loadControllers(oasFile, config);
    return new OASRouter(oasFile, (req, res, _next) => {  
      const requestPath = req.route.path;
      const method = req.method;
      controllers[requestPath][method](req, res);
    });
  }

...
```
2. Modify the call to controllers (line 16) to inject your dependencies inside its arguments:
```js
// call controller
controllers[requestPath][method](req, res, config.di);
```
3. Modify your controllers to add the new argument
```js
export function DIController(req, res, deps) {
    // Do something
}
```
4. Disable router in configuration and use this middleware instead at position 4 (after the response validation, see the docs [here](https://oas-tools.github.io/docs/customization/external-modules))
```jsonc
// .oastoolsrc
{
    ...
    "middleware": {
        "router": {
            "disable": true
        }
    }
}
```

```js
import { NewRouter } from './newRouter.js';

const depsCfg= {
  di: {
     dep1: import('some-dependency',
     dep2: import('another-dependency') 
  }
}

oasTools.use(NewRouter, depsCfg, 4);
oasTools.initialize(app).then(() => ...)
```

I hope you find this workaround useful. There are many other ways to implement DI, like adding your dependencies to the `req` or `res.locals` objects to avoid modifying your controllers. Find the way that works better for you and feel free to leave a sugestion if you find a good implementation. This feature is interesting and I will possibly implement it in future versions 😄

**Reply by koubas** [reply]
tanks! Seems acceptable for now.
